bug: `keccak_hash::H256` does not implement derived `serde` macros

```bash
   Compiling pmt v0.1.0 (/Users/eureka/Code/patriecia/pmt)
error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
    --> pmt/src/inner.rs:33:16
     |
33   |     root_hash: H256,
     |                ^^^^ the trait `Deserialize<'_>` is not implemented for `H256`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               &'a Path
               &'a [u8]
               &'a str
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
             and 137 others
note: required by a bound in `next_element`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1729:12
     |
1727 |     fn next_element<T>(&mut self) -> Result<Option<T>, Self::Error>
     |        ------------ required by a bound in this associated function
1728 |     where
1729 |         T: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `SeqAccess::next_element`

error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
    --> pmt/src/inner.rs:33:16
     |
33   |     root_hash: H256,
     |                ^^^^ the trait `Deserialize<'_>` is not implemented for `H256`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               &'a Path
               &'a [u8]
               &'a str
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
             and 137 others
note: required by a bound in `next_value`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1868:12
     |
1866 |     fn next_value<V>(&mut self) -> Result<V, Self::Error>
     |        ---------- required by a bound in this associated function
1867 |     where
1868 |         V: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `MapAccess::next_value`

error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
  --> pmt/src/inner.rs:32:5
   |
32 | /     /// 32 byte hash of the trie's root node.
33 | |     root_hash: H256,
   | |___________________^ the trait `Deserialize<'_>` is not implemented for `H256`
   |
   = help: the following other types implement trait `Deserialize<'de>`:
             &'a Path
             &'a [u8]
             &'a str
             ()
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
           and 137 others
note: required by a bound in `inner::_::_serde::__private::de::missing_field`
  --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/private/de.rs:22:8
   |
20 | pub fn missing_field<'de, V, E>(field: &'static str) -> Result<V, E>
   |        ------------- required by a bound in this function
21 | where
22 |     V: Deserialize<'de>,
   |        ^^^^^^^^^^^^^^^^ required by this bound in `missing_field`

error[E0277]: the trait bound `H256: Serialize` is not satisfied
    --> pmt/src/inner.rs:26:46
     |
26   |   #[derive(Debug, Clone, Default, Deserialize, Serialize)]
     |                                                ^^^^^^^^^ the trait `Serialize` is not implemented for `H256`
...
32   | /     /// 32 byte hash of the trie's root node.
33   | |     root_hash: H256,
     | |___________________- required by a bound introduced by this call
     |
     = help: the following other types implement trait `Serialize`:
               &'a T
               &'a mut T
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
               (T0, T1, T2, T3, T4, T5)
             and 136 others
note: required by a bound in `inner::_::_serde::ser::SerializeStruct::serialize_field`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/ser/mod.rs:1901:12
     |
1895 |     fn serialize_field<T: ?Sized>(
     |        --------------- required by a bound in this associated function
...
1901 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`

error[E0277]: the trait bound `H256: Serialize` is not satisfied
    --> pmt/src/node.rs:117:24
     |
117  | #[derive(Debug, Clone, Serialize, Deserialize)]
     |                        ^^^^^^^^^ the trait `Serialize` is not implemented for `H256`
118  | pub struct HashNode {
119  |     pub hash: H256,
     |     -------------- required by a bound introduced by this call
     |
     = help: the following other types implement trait `Serialize`:
               &'a T
               &'a mut T
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
               (T0, T1, T2, T3, T4, T5)
             and 137 others
note: required by a bound in `inner::_::_serde::ser::SerializeStruct::serialize_field`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/ser/mod.rs:1901:12
     |
1895 |     fn serialize_field<T: ?Sized>(
     |        --------------- required by a bound in this associated function
...
1901 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`

error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
    --> pmt/src/node.rs:119:15
     |
119  |     pub hash: H256,
     |               ^^^^ the trait `Deserialize<'_>` is not implemented for `H256`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               &'a Path
               &'a [u8]
               &'a str
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
             and 143 others
note: required by a bound in `next_element`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1729:12
     |
1727 |     fn next_element<T>(&mut self) -> Result<Option<T>, Self::Error>
     |        ------------ required by a bound in this associated function
1728 |     where
1729 |         T: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `SeqAccess::next_element`

error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
    --> pmt/src/node.rs:119:15
     |
119  |     pub hash: H256,
     |               ^^^^ the trait `Deserialize<'_>` is not implemented for `H256`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               &'a Path
               &'a [u8]
               &'a str
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
             and 143 others
note: required by a bound in `next_value`
    --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1868:12
     |
1866 |     fn next_value<V>(&mut self) -> Result<V, Self::Error>
     |        ---------- required by a bound in this associated function
1867 |     where
1868 |         V: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `MapAccess::next_value`

error[E0277]: the trait bound `H256: Deserialize<'_>` is not satisfied
   --> pmt/src/node.rs:119:5
    |
119 |     pub hash: H256,
    |     ^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `H256`
    |
    = help: the following other types implement trait `Deserialize<'de>`:
              &'a Path
              &'a [u8]
              &'a str
              ()
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
            and 143 others
note: required by a bound in `inner::_::_serde::__private::de::missing_field`
   --> /Users/eureka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/private/de.rs:22:8
    |
20  | pub fn missing_field<'de, V, E>(field: &'static str) -> Result<V, E>
    |        ------------- required by a bound in this function
21  | where
22  |     V: Deserialize<'de>,
    |        ^^^^^^^^^^^^^^^^ required by this bound in `missing_field`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `pmt` (lib) due to 8 previous errors
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,6 +19,7 @@ thiserror = { workspace = true }
 keccak-hash = { workspace = true }
 parking_lot = { workspace = true }
 hashbrown = { workspace = true }
+fixed-hash = { workspace = true }
 
 [workspace]
 members = [
@@ -29,6 +30,7 @@ members = [
 [workspace.dependencies]
 left-right = "0.11.5"
 keccak-hash = "0.9"
+fixed-hash = "0.7.0"
 log = "0.4.16"
 rlp = "0.5.1"
 rand = "0.8.3"

--- a/jmt/Cargo.toml
+++ b/jmt/Cargo.toml
@@ -12,7 +12,6 @@ edition = "2021"
 
 [lib]
 name = "jmt"
-path = "jmt"
 
 [features]
 default = ["ics23", "std"]

--- a/pmt/Cargo.toml
+++ b/pmt/Cargo.toml
@@ -12,6 +12,7 @@ name = "pmt"
 left-right = { workspace = true }
 hashbrown = { workspace = true }
 keccak-hash = { workspace = true }
+fixed-hash = { workspace = true }
 log = { workspace = true }
 parking_lot = { workspace = true }
 rlp = { workspace = true }

--- a/pmt/Cargo.toml
+++ b/pmt/Cargo.toml
@@ -7,7 +7,6 @@ authors = ["VRRB Labs <info@vrrb.io>"]
 
 [lib]
 name = "pmt"
-path = "pmt"
 
 [dependencies]
 left-right = { workspace = true }

--- a/pmt/src/error.rs
+++ b/pmt/src/error.rs
@@ -1,8 +1,7 @@
-use keccak_hash::H256;
 use rlp::DecoderError;
 use thiserror::Error;
 
-use crate::{nibbles::Nibbles, node::NodeError};
+use crate::{nibbles::Nibbles, node::NodeError, serde_hash::H256};
 
 #[derive(Error, Debug, Clone, Eq, PartialEq)]
 pub enum TrieError {

--- a/pmt/src/inner.rs
+++ b/pmt/src/inner.rs
@@ -6,7 +6,6 @@ use std::sync::Arc;
 use std::{borrow::BorrowMut, fmt::Display};
 
 use hashbrown::{HashMap, HashSet};
-use keccak_hash::{keccak, H256};
 use rlp::{Prototype, Rlp, RlpStream};
 use serde::{Deserialize, Serialize};
 
@@ -17,6 +16,7 @@ use crate::{
     nibbles::Nibbles,
     node::{BranchNode, ExtensionNode, HashNode, Node},
     result::Result,
+    serde_hash::{keccak, H256},
     trie::Trie,
     TrieIterator,
 };
@@ -433,7 +433,7 @@ where
                 ))
             }
             Node::Branch(ref mut branch) => {
-                let mut borrow_branch = branch.borrow_mut();
+                let borrow_branch = branch.borrow_mut();
 
                 if partial.at(0) == 0x10 {
                     borrow_branch.value = Some(value);
@@ -447,7 +447,7 @@ where
                 Ok(Node::Branch(branch.clone()))
             }
             Node::Extension(ext) => {
-                let mut borrow_ext = ext.borrow_mut();
+                let borrow_ext = ext.borrow_mut();
 
                 let prefix = &borrow_ext.prefix;
                 let mut sub_node = borrow_ext.node.clone();
@@ -525,7 +525,7 @@ where
                 Ok((Node::Leaf(leaf.clone()), false))
             }
             Node::Branch(branch) => {
-                let mut borrow_branch = branch.borrow_mut();
+                let borrow_branch = branch.borrow_mut();
 
                 if partial.at(0) == 0x10 {
                     borrow_branch.value = None;

--- a/pmt/src/lib.rs
+++ b/pmt/src/lib.rs
@@ -8,6 +8,7 @@ pub mod trie;
 pub mod trie_iterator;
 
 pub mod common;
+pub mod serde_hash;
 pub use common::*;
 pub use db::*;
 pub use error::*;
@@ -19,7 +20,7 @@ pub use trie_iterator::*;
 pub(crate) mod nibbles;
 pub(crate) mod node;
 
-pub use keccak_hash::H256;
+pub use serde_hash::H256;
 
 #[cfg(test)]
 mod tests {
@@ -28,13 +29,13 @@ mod tests {
         sync::Arc,
     };
 
-    use keccak_hash::{keccak, H256};
     use rand::{distributions::Alphanumeric, seq::SliceRandom, thread_rng, Rng};
 
     use crate::{
         db::{Database, MemoryDB},
         error::TrieError,
         nibbles::Nibbles,
+        serde_hash::{keccak, H256},
     };
     use crate::{InnerTrie, Trie};
 

--- a/pmt/src/node.rs
+++ b/pmt/src/node.rs
@@ -1,8 +1,7 @@
-use keccak_hash::H256;
 use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
-use crate::{common::BRANCHING_FACTOR, nibbles::Nibbles};
+use crate::{common::BRANCHING_FACTOR, nibbles::Nibbles, serde_hash::H256};
 
 pub type Link = Box<Node>;
 

--- a/pmt/src/serde_hash.rs
+++ b/pmt/src/serde_hash.rs
@@ -1,0 +1,24 @@
+//! An adaptation of the original `keccak` function and `H256` type
+//! that handles the implementation of `serde`.
+
+use fixed_hash::construct_fixed_hash;
+use keccak_hash::write_keccak;
+use serde::{Deserialize, Serialize};
+
+construct_fixed_hash! {
+    /// Fixed-size uninterpreted hash type with 32 bytes (256 bits) size.
+    ///
+    /// This type is an adaptation of `keccak::H256` that accounts for the `serde` derive macros
+    /// that are used within the `pmt` portion of the `patriecia` library.
+    #[cfg_attr(feature = "scale-info", derive(TypeInfo))]
+    #[derive(Serialize, Deserialize)]
+    pub struct H256(32);
+}
+
+/// An adaptation of the `keccak_hash::keccak` function that returns
+/// a [serde_hash::H256].
+pub fn keccak<T: AsRef<[u8]>>(s: T) -> crate::serde_hash::H256 {
+    let mut result = [0u8; 32];
+    write_keccak(s, &mut result);
+    crate::serde_hash::H256(result)
+}

--- a/pmt/src/trie.rs
+++ b/pmt/src/trie.rs
@@ -1,8 +1,7 @@
-use keccak_hash::H256;
-
 use crate::common::{Key, OwnedValue, Value};
 use crate::db::Database;
 use crate::result::Result;
+use crate::serde_hash::H256;
 
 pub trait Trie<D: Database> {
     /// Returns the value for key stored in the trie.

